Applications/Spire/Source/Ui/TagBox.cpp

```diff
@@ -0,0 +1,513 @@
+#include "Spire/Ui/TagBox.hpp"
+#include <QHBoxLayout>
+#include <QKeyEvent>
+#include "Spire/Spire/Dimensions.hpp"
+#include "Spire/Ui/ArrayListModel.hpp"
+#include "Spire/Ui/Box.hpp"
+#include "Spire/Ui/CustomQtVariants.hpp"
+#include "Spire/Ui/ListItem.hpp"
+#include "Spire/Ui/ListView.hpp"
+#include "Spire/Ui/Tag.hpp"
+
+using namespace boost;
+using namespace boost::signals2;
+using namespace Spire;
+using namespace Spire::Styles;
+
+namespace {
+  auto INPUT_BOX_STYLE(StyleSheet style) {
+    style.get(Any()).
+      set(vertical_padding(scale_height(3)));
+    return style;
+  }
+
+  auto LIST_ITEM_STYLE() {
+    auto style = StyleSheet();
+    style.get(Any()).
+      set(BackgroundColor(QColor(Qt::transparent))).
+      set(border_size(0)).
+      set(padding(0));
+    return style;
+  }
+
+  auto LIST_VIEW_STYLE(StyleSheet style) {
+    style.get(Any()).
+      set(ListItemGap(scale_width(4))).
+      set(ListOverflowGap(scale_width(3))).
+      set(Overflow::WRAP).
+      set(Qt::Horizontal);
+    auto& list_item_rule = style.get(Any() >> is_a<ListItem>());
+    auto list_item_style = LIST_ITEM_STYLE();
+    for(auto& rule : list_item_style.get_rules()) {
+      for(auto& property : rule.get_block()) {
+        list_item_rule.set(property);
+      }
+    }
+    return style;
+  }
+
+  auto TEXT_BOX_STYLE(StyleSheet style) {
+    style.get(Any() || Hover() || Focus() || Disabled()).
+      set(BackgroundColor(QColor(Qt::transparent))).
+      set(border_size(0)).
+      set(horizontal_padding(0)).
+      set(vertical_padding(scale_height(2)));
+    return style;
+  }
+}
+
+struct TagBox::PartialListModel : public ListModel {
+  std::shared_ptr<ListModel> m_source;
+
+  explicit PartialListModel(std::shared_ptr<ListModel> source)
+    : m_source(std::move(source)) {}
+
+  int get_size() const override {
+    return m_source->get_size() + 2;
+  }
+
+  const std::any& at(int index) const override {
+    if(index < 0 || index >= get_size()) {
+      throw std::out_of_range("The index is out of range.");
+    }
+    if(index < m_source->get_size()) {
+      return m_source->at(index);
+    }
+    static auto value = std::any();
+    return value;
+  }
+
+  QValidator::State set(int index, const std::any& value) override {
+    if(index < 0 || index >= get_size()) {
+      return QValidator::State::Invalid;
+    }
+    if(index < m_source->get_size()) {
+      return m_source->set(index, value);
+    }
+    return QValidator::State::Acceptable;
+  }
+
+  connection connect_operation_signal(
+      const OperationSignal::slot_type& slot) const override {
+    return m_source->connect_operation_signal(slot);
```

Comment:
This only works if ```PartialListModel``` owns ```m_source```. If it doesn't own ```m_source``` then you can not pass ```slot``` directly to ```m_source``` since when ```PartialListModel``` is destroyed, all those slots needs to be destroyed as well.

If ```PartialListModel``` owns ```m_source``` then there is no issue.